### ISSUE TITLE
Using roles defined in Azure AD

### ISSUE THREAD

**Issue by robertfshort** [Badgerati/Pode#887]
I've got Azure AD authentication working, and in Azure I setup several app roles. But when I follow the tutorial for Azure AD authentication I don't seem to get any of the information about those roles.

Potentially are there more params here that are not documented in the tutorial?
```
$scheme | Add-PodeAuth -Name 'Login' -FailureUrl '/login' -SuccessUrl '/' -ScriptBlock {
        param($user, $accessToken, $refreshToken)
        write-podehost $user
        # check if the user is valid

        return @{ User = $user }
    }
```
Or perhaps there is some way to get the complete json response from Azure AD?

**Comment by robertfshort** [comment]
I just noticed that the AUD in the token isn't actually my app, but MS graph instead... How do I get it to return info from my defined Azure AD app?

**Comment by Badgerati** [comment]
Hi @robertfshort,

Are you seeing the docs from [here on GitHub](https://badgerati.github.io/Pode/Tutorials/Authentication/Methods/AzureAD/)? On there you'll see there's a 4th `$response` parameter for the `Add-PodeAuth` scriptblock 😄 (the docs are somehow hosted in other places, that are usually out-of-date or missing pages)

The `$response` is the raw result from Azure AD's OAuth, and should hopefully contain what you're looking for. If it's in the `id_token` property, you can use `ConvertFrom-PodeJwt -Token $response.id_token -IgnoreSignature` to parse the JWT.

Hope that helps!

**Comment by robertfshort** [comment]
I figured out a convoluted way of getting the roles... but it's not pretty. Essentially, using the token to get the UPN, I need to call 
```
"https://graph.microsoft.com/beta/users/{0}/appRoleAssignments" -f $upn
```
then, I need to know what all of the unique ID's are for all the app roles on the app and see if they are in the list rather than using the name of the role (which precludes using a good naming system for roles where you could do partial matches (like if name -contains 'write' then allow this)

Basically it would be much easier if the access token had the app roles for the proper client id in it, but after spending hours trying to trace how PODE makes those calls to the Azure AD page I can confidently say I have no clue how it's done in PODE...

**Comment by robertfshort** [comment]
If I try ConvertFrom-PodeJwt -Token $response.id_token -IgnoreSignature it just makes the page reload constantly. The $reponse variable when added to the params appears to be empty when I try to write it's contents to the pode host.

**Comment by robertfshort** [comment]
Just for further comment, every time I try to use the pode function to convert from JWT it forces the page to reload continuously...

I've been using this function I found elsewhere:
```
function Parse-JWTtoken {
 
        [cmdletbinding()]
        param([Parameter(Mandatory=$true)][string]$token)
     
        #Validate as per https://tools.ietf.org/html/rfc7519
        #Access and ID tokens are fine, Refresh tokens will not work
        if (!$token.Contains(".") -or !$token.StartsWith("eyJ")) { Write-Error "Invalid token" -ErrorAction Stop }
     
        #Header
        $tokenheader = $token.Split(".")[0].Replace('-', '+').Replace('_', '/')
        #Fix padding as needed, keep adding "=" until string length modulus 4 reaches 0
        while ($tokenheader.Length % 4) { Write-Verbose "Invalid length for a Base-64 char array or string, adding ="; $tokenheader += "=" }
        Write-Verbose "Base64 encoded (padded) header:"
        Write-Verbose $tokenheader
        #Convert from Base64 encoded string to PSObject all at once
        Write-Verbose "Decoded header:"
        #[System.Text.Encoding]::ASCII.GetString([system.convert]::FromBase64String($tokenheader)) | ConvertFrom-Json | fl | Out-Default
     
        #Payload
        $tokenPayload = $token.Split(".")[1].Replace('-', '+').Replace('_', '/')
        #Fix padding as needed, keep adding "=" until string length modulus 4 reaches 0
        while ($tokenPayload.Length % 4) { Write-Verbose "Invalid length for a Base-64 char array or string, adding ="; $tokenPayload += "=" }
        Write-Verbose "Base64 encoded (padded) payoad:"
        Write-Verbose $tokenPayload
        #Convert to Byte array
        $tokenByteArray = [System.Convert]::FromBase64String($tokenPayload)
        #Convert to string array
        $tokenArray = [System.Text.Encoding]::ASCII.GetString($tokenByteArray)
        Write-Verbose "Decoded array in JSON format:"
        Write-Verbose $tokenArray
        #Convert from JSON to PSObject
        $tokobj = $tokenArray | ConvertFrom-Json
        Write-Verbose "Decoded Payload:"
        
        return $tokobj
    }
```

**Comment by Badgerati** [comment]
Hi @robertfshort,

When Pode makes the calls for Azure AD it does the following:

1. It redirects the user to `https://login.microsoftonline.com/<tentantId>/oauth2/v2.0/authorize`
2. Once authenticated, Azure redirects the user to a callback route in Pode, with a `?code=` query parameter
3. Pode uses the `code` to call `https://login.microsoftonline.com/<tenantId>/oauth2/v2.0/token` for an access token
4. Pode uses that token to call `https://graph.microsoft.com/oidc/userinfo` to get information about the user
5. Pode passes the user, access_token, refresh_token, and raw token call response to your `Add-PodeAuth`

If the `$response` is empty, which version of Pode are you using? the `$response` being sent is new, as of v2.5.0, plus `ConvertFrom-PodeJwt`. It's likely reloading continuously because the function doesn't exist 🤔 

If you are on v2.5.0, try `$response | out-default` instead, and wrap everything in a try/catch:
```powershell
$scheme | Add-PodeAuth -Name 'Login' -FailureUrl '/login' -SuccessUrl '/' -ScriptBlock {
    param($user, $accessToken, $refreshToken, $response)
    try {
        $response | out-default
        ConvertFrom-PodeJwt -Token $response.id_token -IgnoreSignature
    }
    catch {
        $_.Exception | out-default
    }
    return @{ User = $user}
}
```

That should catch the error if there is one, so we can see what needs fixing.

**Comment by robertfshort** [comment]
Apparently I had 2.4.2, yes the ID token has the roles I am looking for.

**Comment by robertfshort** [comment]
One more questions about the flow here... when is that ID_token generated? Step 2 in your example above?

**Comment by Badgerati** [comment]
It'd be at step 3 :)

The response from calling the Token URL contains the id_token, access_token, refresh_token, and more. It's the raw `$response` available in `Add-PodeAuth`.
